File: LeetCode/LeetCode/remove_end.cc
#include<iostream>

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x):val(x), next(NULL) {}
};

class Solution {
    public:
        ListNode *removeNthFromEnd(ListNode *head, int n) {
            ListNode dummy{0};
            dummy.next = head;
            auto slow = &dummy;
            auto fast = &dummy;

            while( n > 0) {
                fast = fast ->next;
                --n;
            }

            while(fast->next != nullptr) {
                slow = slow->next;
                fast = fast->next;
            }

            auto node_to_delete = slow->next;
            slow->next = slow->next->next;
            delete node_to_delete;

            return dummy.next;
        }
};

int main(int argc, char **argv)
{
	int i;
	struct ListNode *list, *p, *prev = NULL;
	for (i = 1; i < 5; i++) {
		p = new ListNode(i); 
		if (i == 1) {
			list = p;
		}
		if (prev != NULL) {
			prev->next = p;
		}
		prev = p;
	}
	
    Solution s;
	list = s.removeNthFromEnd(list,2);
	if (list != NULL) {
		for (p = list; p != NULL; p = p->next) {
			printf("%d\n", p->val);
		}
	}
	return 0;
}

File: LeetCode/LeetCode/remove_end.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

    def __repr__(self):
        if self is None:
            return "Nil"
        else:
            return "{} -> {}".format(self.val, repr(self.next))

class Solution:
    def removeNthFromEnd(self,head, n):
        dummy = ListNode(-1)
        dummy.next = head
        slow, fast = dummy, dummy

        for i in xrange(n):
            fast = fast.next

        while fast.next:
            slow, fast = slow.next, fast.next

        slow.next = slow.next.next

        return dummy.next

if __name__ == "__main__":
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next  = ListNode(3)
    head.next.next.next  = ListNode(4)

    print(Solution().removeNthFromEnd(head, 2))
